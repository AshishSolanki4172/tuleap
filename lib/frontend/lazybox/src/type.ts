/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
