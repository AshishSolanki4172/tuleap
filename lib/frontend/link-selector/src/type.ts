/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
