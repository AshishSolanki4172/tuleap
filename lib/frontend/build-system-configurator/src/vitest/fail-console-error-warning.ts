/**
 * Copyright (c) Enalean, 2022-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { beforeEach, afterEach, vi } from "vitest";

let is_console_error_or_warning = false;

beforeEach(() => {
    is_console_error_or_warning = false;
    const originalError = global.console.error;
    vi.spyOn(global.console, "error").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalError(...args);
    });
    const originalWarn = global.console.warn;
    vi.spyOn(global.console, "warn").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalWarn(...args);
    });
});

afterEach(() => {
    if (is_console_error_or_warning) {
        throw new Error("Console warnings and errors are not allowed");
    }
});
