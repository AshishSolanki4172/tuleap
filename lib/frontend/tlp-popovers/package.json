{
  "author": "Enalean Team",
  "name": "@tuleap/tlp-popovers",
  "homepage": "https://tuleap.org",
  "license": "GPL-2.0-or-later",
  "private": true,
  "version": "0.0.0",
  "module": "dist/tlp-popovers.es.js",
  "main": "dist/tlp-popovers.umd.js",
  "style": "dist/style.css",
  "dependencies": {
    "@floating-ui/dom": "^0.4.2"
  },
  "scripts": {
    "typecheck": "tsc --noEmit",
    "build": "vite build",
    "watch": "vite build --watch --mode development --minify false",
    "test": "jest"
  },
  "devDependencies": {
    "@tuleap/build-system-configurator": "link:../build-system-configurator"
  }
}
