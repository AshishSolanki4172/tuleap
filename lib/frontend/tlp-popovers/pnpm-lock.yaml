lockfileVersion: 5.3

specifiers:
  '@floating-ui/dom': ^0.4.2
  '@tuleap/build-system-configurator': link:../build-system-configurator

dependencies:
  '@floating-ui/dom': 0.4.4

devDependencies:
  '@tuleap/build-system-configurator': link:../build-system-configurator

packages:

  /@floating-ui/core/0.6.1:
    resolution: {integrity: sha512-Y30eVMcZva8o84c0HcXAtDO4BEzPJMvF6+B7x7urL2xbAqVsGJhojOyHLaoQHQYjb6OkqRq5kO+zeySycQwKqg==}
    dev: false

  /@floating-ui/dom/0.4.4:
    resolution: {integrity: sha512-0Ulu3B/dqQplUUSqnTx0foSrlYuMN+GTtlJWvNJwt6Fr7/PqmlR/Y08o6/+bxDWr6p3roBJRaQ51MDZsNmEhhw==}
    dependencies:
      '@floating-ui/core': 0.6.1
    dev: false
