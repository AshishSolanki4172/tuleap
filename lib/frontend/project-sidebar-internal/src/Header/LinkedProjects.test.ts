/**
 * Copyright (c) 2022-Present Enalean
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { describe, it, expect, vi } from "vitest";
import { shallowMount } from "@vue/test-utils";
import LinkedProjects from "./LinkedProjects.vue";
import { example_config } from "../project-sidebar-example-config";
import { ref } from "vue";
import * as tlp_popovers from "@tuleap/tlp-popovers";
import type { Popover } from "@tuleap/tlp-popovers";
import * as strict_inject from "@tuleap/vue-strict-inject";

vi.mock("@tuleap/vue-strict-inject");

describe("LinkedProjects", () => {
    it("displays the linked projects with a popover", () => {
        const create_popover_spy = vi
            .spyOn(tlp_popovers, "createPopover")
            .mockReturnValue({} as Popover);

        vi.spyOn(strict_inject, "strictInject").mockReturnValue(ref(example_config));

        const wrapper = shallowMount(LinkedProjects);

        expect(wrapper.element).toMatchSnapshot();
        expect(create_popover_spy).toHaveBeenCalled();
    });

    it("display nothing if there is no linked projects", () => {
        const config = example_config;
        config.project.linked_projects = null;
        vi.spyOn(strict_inject, "strictInject").mockReturnValue(ref(config));
        const wrapper = shallowMount(LinkedProjects);

        expect(wrapper.element.textContent).toBe("");
    });

    it.each([
        [5, true, false],
        [6, false, true],
    ])(
        `Given the config does not specify the nb max project to display in sidebar, it defaults to 5.
     When nb projects = %s,
     Then projects in sidebar are displayed = %s
     And popover is displayed = %s`,
        (nb, expected_in_sidebar, expected_popover) => {
            vi.spyOn(tlp_popovers, "createPopover").mockReturnValue({} as Popover);

            const projects = Array(nb).fill({ name: "acme" });

            const config = {
                ...example_config,
                project: {
                    ...example_config.project,
                    linked_projects: {
                        ...example_config.project.linked_projects,
                        projects,
                    },
                },
            };
            vi.spyOn(strict_inject, "strictInject").mockReturnValue(ref(config));

            const wrapper = shallowMount(LinkedProjects);

            expect(wrapper.element).toMatchSnapshot();
            expect(wrapper.find("[data-test=nav-bar-linked-projects]").exists()).toBe(
                expected_in_sidebar,
            );
            expect(
                wrapper
                    .find("[data-test=popover]")
                    .classes("project-sidebar-linked-projects-popover-nb-max-exceeded"),
            ).toBe(expected_popover);
        },
    );
});
