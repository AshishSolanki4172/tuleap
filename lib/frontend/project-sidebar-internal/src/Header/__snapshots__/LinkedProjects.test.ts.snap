// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinkedProjects > Given the config does not specify the nb max project to display in sidebar, it defaults to 5.
     When nb projects = 5,
     Then projects in sidebar are displayed = true
     And popover is displayed = false 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="project-sidebar-linked-projects"
    id="project-sidebar-linked-projects"
  >
    <span
      class="project-sidebar-linked-projects-title"
    >
      <span
        class="project-sidebar-linked-projects-icon"
      >
        <i
          aria-hidden="true"
          class="fa-solid fa-fw fa-turn-up fa-flip-horizontal project-sidebar-linked-projects-icon-parent"
        />
      </span>
      <span
        class="project-sidebar-linked-projects-label"
      />
    </span>
    <ul
      class="project-sidebar-linked-projects-list"
      data-test="nav-bar-linked-projects"
    >
      
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            acme
          </span>
        </a>
      </li>
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            acme
          </span>
        </a>
      </li>
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            acme
          </span>
        </a>
      </li>
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            acme
          </span>
        </a>
      </li>
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            acme
          </span>
        </a>
      </li>
      
    </ul>
  </div>
  <div
    class="tlp-popover project-sidebar-linked-projects-popover"
    data-test="popover"
    id="project-sidebar-linked-projects-popover"
  >
    <div
      class="tlp-popover-arrow project-sidebar-linked-projects-popover-arrow"
    />
    <div
      class="tlp-popover-header"
    >
      <i
        aria-hidden="true"
        class="fa-solid fa-fw fa-turn-up fa-flip-horizontal project-sidebar-linked-projects-icon-parent"
      />
      <h1
        class="tlp-popover-title"
      />
    </div>
    <div
      class="tlp-popover-body project-sidebar-linked-projects-popover-content"
    >
      <ul
        class="project-sidebar-linked-projects-list-popover"
      >
        
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        
      </ul>
    </div>
  </div>
  
</div>
`;

exports[`LinkedProjects > Given the config does not specify the nb max project to display in sidebar, it defaults to 5.
     When nb projects = 6,
     Then projects in sidebar are displayed = false
     And popover is displayed = true 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="project-sidebar-linked-projects"
    id="project-sidebar-linked-projects"
  >
    <span
      class="project-sidebar-linked-projects-title"
    >
      <span
        class="project-sidebar-linked-projects-icon"
      >
        <i
          aria-hidden="true"
          class="fa-solid fa-fw fa-turn-up fa-flip-horizontal project-sidebar-linked-projects-icon-parent"
        />
      </span>
      <span
        class="project-sidebar-linked-projects-label"
      />
    </span>
    <!--v-if-->
  </div>
  <div
    class="tlp-popover project-sidebar-linked-projects-popover project-sidebar-linked-projects-popover-nb-max-exceeded"
    data-test="popover"
    id="project-sidebar-linked-projects-popover"
  >
    <div
      class="tlp-popover-arrow project-sidebar-linked-projects-popover-arrow"
    />
    <div
      class="tlp-popover-header"
    >
      <i
        aria-hidden="true"
        class="fa-solid fa-fw fa-turn-up fa-flip-horizontal project-sidebar-linked-projects-icon-parent"
      />
      <h1
        class="tlp-popover-title"
      />
    </div>
    <div
      class="tlp-popover-body project-sidebar-linked-projects-popover-content"
    >
      <ul
        class="project-sidebar-linked-projects-list-popover"
      >
        
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              acme
            </span>
          </div>
        </li>
        
      </ul>
    </div>
  </div>
  
</div>
`;

exports[`LinkedProjects > displays the linked projects with a popover 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="project-sidebar-linked-projects"
    id="project-sidebar-linked-projects"
  >
    <span
      class="project-sidebar-linked-projects-title"
    >
      <span
        class="project-sidebar-linked-projects-icon"
      >
        <i
          aria-hidden="true"
          class="fa-solid fa-fw fa-tlp-project-boxes"
        />
      </span>
      <span
        class="project-sidebar-linked-projects-label"
      >
        2 aggregated projects
      </span>
    </span>
    <ul
      class="project-sidebar-linked-projects-list"
      data-test="nav-bar-linked-projects"
    >
      
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
          href="https://myinstance.example.com/projects/teama"
        >
          <i
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon fa-solid fa-box-archive"
          />
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            Team A
          </span>
        </a>
      </li>
      <li
        class="project-sidebar-linked-projects-item"
      >
        <a
          class="project-sidebar-linked-projects-item-link"
          href="https://myinstance.example.com/projects/teamb"
        >
          <span
            aria-hidden="true"
            class="project-sidebar-linked-projects-item-icon"
          >
            🏡
          </span>
          <span
            class="project-sidebar-linked-projects-item-name"
          >
            Team B
          </span>
        </a>
      </li>
      
    </ul>
  </div>
  <div
    class="tlp-popover project-sidebar-linked-projects-popover"
    data-test="popover"
    id="project-sidebar-linked-projects-popover"
  >
    <div
      class="tlp-popover-arrow project-sidebar-linked-projects-popover-arrow"
    />
    <div
      class="tlp-popover-header"
    >
      <i
        aria-hidden="true"
        class="fa-solid fa-fw fa-tlp-project-boxes"
      />
      <h1
        class="tlp-popover-title"
      >
        2 aggregated projects
      </h1>
    </div>
    <div
      class="tlp-popover-body project-sidebar-linked-projects-popover-content"
    >
      <ul
        class="project-sidebar-linked-projects-list-popover"
      >
        
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <i
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon fa-solid fa-box-archive"
            />
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              Team A
            </span>
          </div>
        </li>
        <li
          class="project-sidebar-linked-projects-item"
        >
          <div
            class="project-sidebar-linked-projects-item-link"
          >
            <span
              aria-hidden="true"
              class="project-sidebar-linked-projects-item-icon"
            >
              🏡
            </span>
            <span
              class="project-sidebar-linked-projects-item-name"
            >
              Team B
            </span>
          </div>
        </li>
        
      </ul>
    </div>
  </div>
  
</div>
`;
